### PR TITLE
 [Security] always check the token on non-lazy firewalls 

### LINKED ISSUE #34357: [Security] Potential BC break with guard::start() returning nothing
**Symfony version(s) affected**: 4.4-BETA2

**Description**  

I updated an application to 4.4-BETA2 and I have a failing test (only one!).
I request the API entrypoint without credential, and I assert it returns a 401.
With SF 4.3 => OK
With SF 4.4 => KO, I got a 200

**How to reproduce**  

I created 2 new projets with the website skeleton. One on 4.4 and one on 4.3 and I create an empty guard + a controller (thanks to maker bundle). Then I updated the security to:
```yaml
security:
    firewalls:
        main:
            pattern: ^/
            stateless: true
            anonymous: false
            guard:
                authenticators:
                    - App\Security\AppCustomAuthenticator
```

So I'm expecting a 401 when no credential are present in the request because of `anonymous: false`.

With Symfony 4.4 => 200
With Symfony 4.3 => 500 : `The App\Security\AppCustomAuthenticator::start() method must return a Response object (NULL returned)`

**Possible Solution**  

I'm not sure my application has the same side effect, I will dig. 
But anyway, with anonymous=false, I expect to throw a 401 when the credentials are not present (and so the authentication is not done)

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #34357
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by stof** [comment]
isn't this breaking the support for lazy firewalls by always using the token, even when you try to disable checks for some URL patterns ?

**Comment by lyrixx** [comment]
I really don't know. I open the PR to get some feedbacks...

**Comment by nicolas-grekas** [comment]
Yes, it breaks lazyness. Can you please add a test case that reproduces the situation you have?

**Comment by lyrixx** [comment]
1. I reverted my commit to show the bug
2. I added a new test case to ... show the bug :)

**Comment by stof** [comment]
@nicolas-grekas maybe lazyness should be enabled only for firewalls enabling anonymous access (as such firewalls are expected to be able to deal with cases where the user might not be forced to be actually authenticated already). The lazyness indeed breaks the previous assumption that a firewall will force to have an authenticated token on all requests even when no check is performed. And when authenticated tokens can only be created for an authenticated user and not for anonymous users, the previous implementation was implicitly adding a check on all requests.

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/AnoynousTest.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\Tests\Functional;
+
+final class AnoynousTest extends AbstractWebTestCase
```
Comment (derrabus) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/AnoynousTest.php]:
```suggestion
final class AnonymousTest extends AbstractWebTestCase
```

**Comment by nicolas-grekas** [comment]
@stof implemented, can you please have a look?

@lyrixx I push-forced on your fork - can you also please have a look at the reason why the test is failing now?

**Comment by nicolas-grekas** [comment]
> can you also please have a look at the reason why the test is failing now?

I figured out on my own :)

PR ready.

**Review by lyrixx — COMMENTED** [review]
:+1: Simple, like it

**Review by chalasr — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @lyrixx.

**Comment by lyrixx** [comment]
@nicolas-grekas Thanks you for finish it

**Pull request merged**